server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java

```diff
@@ -306,4 +307,23 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
       Param.SORT, METRIC_SORT, METRIC_PERIOD_SORT, PARAM_METRIC_SORT);
     return componentTreeWsRequest;
   }
+
+  private static WsMeasures.Component.Builder toWsComponent(ComponentDto component, Map<MetricDto, ComponentTreeData.Measure> measures,
+    Map<String, ComponentDto> referenceComponentsByUuid) {
+    WsMeasures.Component.Builder wsComponent = componentDtoToWsComponent(component);
+    ComponentDto referenceComponent = referenceComponentsByUuid.get(component.getCopyResourceUuid());
+    if (referenceComponent != null) {
+      wsComponent.setRefId(referenceComponent.uuid());
+      wsComponent.setRefKey(referenceComponent.key());
+    }
+    for (Map.Entry<MetricDto, ComponentTreeData.Measure> entry : measures.entrySet()) {
```

Comment:
This loop should be optimized by reusing the same instances of protobuf Measure.Builder and PeriodValue.Builder (by calling `clear()` before each call). 
Moreover `value` and `variation` must not be autoboxed again to Double. 

```

```